contracts/multihop/src/tests/swap.rs

```diff
@@ -3,158 +3,360 @@ use crate::storage::Swap;
+    assert_eq!(
+        token4.balance(&recipient),
+        4_956i128,
+        "token4 not as expected"
+    );
+}
+
+#[test]
+fn swap_three_different_pools_with_fees() {
+    let env = Env::default();
+
+    let admin = Address::random(&env);
+
+    env.mock_all_auths();
+    env.budget().reset_unlimited();
+
+    let token1 = deploy_and_mint_tokens(&env, &admin, 10_000_000i128);
+    let token2 = deploy_and_mint_tokens(&env, &admin, 10_000_000i128);
+    let token3 = deploy_and_mint_tokens(&env, &admin, 10_000_000i128);
+    let token4 = deploy_and_mint_tokens(&env, &admin, 10_000_000i128);
+
+    // 1. deploy factory
+    let factory_client = deploy_and_initialize_factory(&env, admin.clone());
+
+    deploy_and_initialize_lp(
+        &env,
+        &factory_client,
+        admin.clone(),
+        token1.address.clone(),
+        1_000_000,
+        token2.address.clone(),
+        1_000_000,
+        Some(1_000),
+    );
+    deploy_and_initialize_lp(
+        &env,
+        &factory_client,
+        admin.clone(),
+        token2.address.clone(),
+        2_000_000,
+        token3.address.clone(),
+        2_000_000,
+        Some(1_000),
+    );
+    deploy_and_initialize_lp(
+        &env,
+        &factory_client,
+        admin.clone(),
+        token3.address.clone(),
+        3_000_000,
+        token4.address.clone(),
+        3_000_000,
+        Some(1_000),
+    );
 
     // 4. swap with multihop
-    let multihop = deploy_multihop_contract(&env, admin, &factory_client.address);
+    let multihop = deploy_multihop_contract(&env, admin.clone(), &factory_client.address);
     let recipient = Address::random(&env);
-    tokens[0].mint(&recipient, &50i128);
-    assert_eq!(tokens[0].balance(&recipient), 50i128);
-    assert_eq!(tokens[3].balance(&recipient), 0i128);
+    token1.mint(&recipient, &10_000i128);
+    assert_eq!(token1.balance(&recipient), 10_000i128);
 
     let swap1 = Swap {
-        offer_asset: tokens[0].address.clone(),
-        ask_asset: tokens[1].address.clone(),
+        offer_asset: token1.address.clone(),
+        ask_asset: token2.address.clone(),
     };
     let swap2 = Swap {
-        offer_asset: tokens[1].address.clone(),
-        ask_asset: tokens[2].address.clone(),
+        offer_asset: token2.address.clone(),
+        ask_asset: token3.address.clone(),
     };
     let swap3 = Swap {
-        offer_asset: tokens[2].address.clone(),
-        ask_asset: tokens[3].address.clone(),
+        offer_asset: token3.address.clone(),
+        ask_asset: token4.address.clone(),
     };
 
     let operations = vec![&env, swap1, swap2, swap3];
 
-    multihop.swap(&recipient, &operations, &50i128);
+    multihop.swap(&recipient, &operations, &10_000i128);
 
-    // 5. check if it goes according to plan
-    assert_eq!(tokens[0].balance(&recipient), 0i128);
-    assert_eq!(tokens[3].balance(&recipient), 50i128);
+    // we start swapping 10_000 tokens
+
+    // token1 => token2
```

Comment:
add another set of assertions that verifies the amount of tokens in between swaps in addition to the comments